### PR TITLE
Fix text.rich to merge widget span

### LINKED ISSUE #111931: `Text.rich` does not compose its semantics label in the correct order for embedded widgets
b/247794358

## Steps to Reproduce

The rich text widget doesn't seem to compose the semantics labels in correct order. It seems to concatenate all the `TextSpan`s, replacing an embedded `WidgetSpan` with u+fffc (unicode object replacement character), and concatenates the `WidgetSpan` label at the end. Here's a test reproducing the issue:

```dart
testWidgets('semantics label not in order', (tester) async {
  await tester.pumpWidget(
    Directionality(
      textDirection: TextDirection.ltr,
      child: Text.rich(
        TextSpan(
          children: [
            TextSpan(text: 'before '),
            WidgetSpan(
              alignment: PlaceholderAlignment.baseline,
              baseline: TextBaseline.alphabetic,
              child: Semantics(label: 'foo'),
            ),
            TextSpan(text: ' after'),
          ],
        ),
      ),
    ),
  );
  expect(tester.getSemantics(find.byType(Text)),
      matchesSemantics(label: 'before foo after'));
});
```

Error message:
```
Expected: has semantics with label: before foo after without actions:
  <content removed>
  Actual: SemanticsNode:<SemanticsNode#1(Rect.fromLTRB(0.0, 0.0, 1591.0, 893.0), label: "before ￼
after\nfoo", textDirection: ltr)>
   Which: label was: before ￼ after
          foo
```

**Expected:** The test should pass. Specifically, the semantic label should be "before foo after".

**Actual:** It's computing the semantic string as "before ￼ after\nfoo". In between "before " and " after", there is a u+fffc unicode character (object replacement character), which is not meaningful to screen readers.

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/111931

This pr let text widget to use child semanticsconfigurations delegate API to add synthetic semantics configurations to be merge up.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chunhtai** [comment]
This is ready for another look.

I can't find a good way to complete replace assembleSemanticsNode in RenderParahraph because, there isn't a good way to generate semantics node from synthetic SemanticsConfiguration with two reasons.

1. Semantics node is cached in RenderObject._semantics. This is important because we want to reuse semantics id as much as possible. This becomes tricky now the semantics node generated from synthetic SemanticsConfiguration need to be have a stable id somehow
2. Needs to figure out a way to set the semantics node rect and transform. Regular semantics node gets its geometry from renderobject directly. The synthetic SemanticsConfiguration need to process these information somehow.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 1]
```diff
@@ -1053,12 +1055,115 @@ class RenderParagraph extends RenderBox
...
+              if (seenTextSpan) {
+                final SemanticsConfiguration textSpanConfig = SemanticsConfiguration();
+                textSpanConfig.textDirection = textDirection;
+                textSpanConfig.attributedLabel = cachedStrings[textSpanLabelIndex];
+                builder.markAsMergeUp(textSpanConfig);
+                textSpanLabelIndex += 1;
+                seenTextSpan = false;
+              }
+              while (childConfigsIndex < childConfigs.length &&
+                     childConfigs[childConfigsIndex].tagsChildrenWith(PlaceholderSpanIndexSemanticsTag(placeholderIndex))) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/paragraph.dart:1091]: I thought the `placeholderIndex` tag is supposed to be unique?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/paragraph.dart:1091]: Ah ok i see now it could be applied on multiple render objects.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 2]
```diff
@@ -1053,12 +1055,115 @@ class RenderParagraph extends RenderBox
...
+      config.childConfigurationsDelegate = (List<SemanticsConfiguration> childConfigs) {
+        final List<AttributedString> cachedStrings = _cachedAttributedLabels ??=  <AttributedString>[];
+        final ChildSemanticsConfigurationsResultBuilder builder = ChildSemanticsConfigurationsResultBuilder();
+        int placeholderIndex = 0;
+        int childConfigsIndex = 0;
+        if (cachedStrings.isNotEmpty) {
+          int textSpanLabelIndex = 0;
+          bool seenTextSpan = false;
+          for (final InlineSpanSemanticsInformation info in _semanticsInfo!) {
+            if (info.isPlaceholder) {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/paragraph.dart:1081]: yes, this will ignore any other placeholder subclass, since we don't know what how they would contribute to the semantics...

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 3]
```diff
@@ -1053,12 +1055,115 @@ class RenderParagraph extends RenderBox
...
+            }(),
+            'all cached strings must be used',
+          );
+
+          if (seenTextSpan) {
+            final SemanticsConfiguration textSpanConfig = SemanticsConfiguration();
+            textSpanConfig.attributedLabel = cachedStrings.last;
+            builder.markAsMergeUp(textSpanConfig);
+          }
+        } else {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/paragraph.dart]: done

**Review by LongCatIsLooong — APPROVED** [review]

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -4295,6 +4312,48 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
   }
 }
 
+/// An fragment that only carries partial information that must not form a
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:4315]:
```suggestion
/// A fragment that only carries partial information that must not form a
```

File: packages/flutter/lib/src/rendering/object.dart [hunk 5]
```diff
@@ -4295,6 +4312,48 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
...
+///
+/// This fragment is generated from synthetic SemanticsConfiguration returned from
+/// [SemanticsConfiguration.childConfigurationsDelegate].
+class _IncompleteSemanticsFragment extends _InterestingSemanticsFragment {
+  _IncompleteSemanticsFragment({
+    required this.config,
+    required super.owner,
+  }) : super(dropsSemanticsOfPreviousSiblings: false);
+
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:4327]: nit: remove extra empty line

File: packages/flutter/lib/src/rendering/object.dart [hunk 6]
```diff
@@ -4295,6 +4312,48 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
...
+  @override
+  void addAll(Iterable<_InterestingSemanticsFragment> fragments) {
+    throw FlutterError('This fragment must be a leaf node');
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:4330]: Any reason this is not wrapped in an assert like markAsExplicit?

File: packages/flutter/lib/src/rendering/object.dart [hunk 7]
```diff
@@ -4295,6 +4312,48 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
...
+  }
+
+  @override
+  void compileChildren({
+    required Rect? parentSemanticsClipRect,
+    required Rect? parentPaintClipRect,
+    required double elevationAdjustment,
+    required List<SemanticsNode> result,
+    required List<SemanticsNode> siblingNodes}) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:4339]: Formatting:
```suggestion
    required List<SemanticsNode> siblingNodes,
  }) {
```

File: packages/flutter/test/widgets/text_test.dart [hunk 8]
```diff
@@ -310,6 +310,152 @@ void main() {
...
+                child: Semantics(label: 'foo'),
+              ),
+              const TextSpan(text: ' after'),
+            ],
+          ),
+        ),
+      ),
+    );
+    expect(tester.getSemantics(find.byType(Text)),
+        matchesSemantics(label: 'before \nfoo\n after'));
```
Comment (goderbauer) [inline packages/flutter/test/widgets/text_test.dart:333]: here and elsewhere in this file: either leave this on one line or format as follows:

```suggestion
    expect(
      tester.getSemantics(find.byType(Text)),
      matchesSemantics(label: 'before \nfoo\n after'),
    );
```

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 9]
```diff
@@ -1076,13 +1089,87 @@ class RenderParagraph extends RenderBox
...
+        textSpanConfig.attributedLabel = AttributedString(buffer.toString(), attributes: attributes!);
+        cachedStrings.add(textSpanConfig.attributedLabel);
+      }
+      builder.markAsMergeUp(textSpanConfig);
+    }
+    return builder.build();
+  }
+
+
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1172]: nit remove two extra empty lines?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 10]
```diff
@@ -1076,13 +1089,87 @@ class RenderParagraph extends RenderBox
...
-        _cachedAttributedLabel = AttributedString(buffer.toString(), attributes: attributes);
+        _cachedAttributedLabels = <AttributedString>[AttributedString(buffer.toString(), attributes: attributes)];
       }
-      config.attributedLabel = _cachedAttributedLabel!;
+      config.attributedLabel = _cachedAttributedLabels![0];
       config.textDirection = textDirection;
     }
   }
 
+  ChildSemanticsConfigurationsResult _childSemanticsConfigurationsDelegate(List<SemanticsConfiguration> childConfigs) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1099]: The logic in this method is hard to follow and reason about with its many levels of nesting and multiple loops some of which short-circuit with `continue`. Any chance this can be refactored a bit (maybe broken out into individual methods or - if all else fails - with some comments to make this more maintainable?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 11]
```diff
@@ -1076,13 +1089,87 @@ class RenderParagraph extends RenderBox
...
+      if (!hasCache) {
+        buffer ??= StringBuffer();
+        attributes ??= <StringAttribute>[];
+        final String label = info.semanticsLabel ?? info.text;
+        for (final StringAttribute infoAttribute in info.stringAttributes) {
+          final TextRange originalRange = infoAttribute.range;
+          attributes.add(
+            infoAttribute.copy(
+              range: TextRange(start: offset + originalRange.start,
+                  end: offset + originalRange.end),
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1148]: formatting

```suggestion
              range: TextRange(
                start: offset + originalRange.start,
                end: offset + originalRange.end,
              ),
```

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 12]
```diff
@@ -1076,13 +1089,87 @@ class RenderParagraph extends RenderBox
...
+        while (childConfigsIndex < childConfigs.length &&
+            childConfigs[childConfigsIndex].tagsChildrenWith(PlaceholderSpanIndexSemanticsTag(placeholderIndex))) {
+          builder.markAsMergeUp(childConfigs[childConfigsIndex]);
+          childConfigsIndex += 1;
+        }
+        placeholderIndex += 1;
+        continue;
+      }
+      seenTextSpan = true;
+      if (!hasCache) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1139]: I may not fully understand this logic, yet, but I would have expected that this branch would put something in the cache? And then there would be an else branch pulling it out of the cache again?

File: packages/flutter/lib/src/rendering/object.dart [hunk 13]
```diff
@@ -4281,6 +4298,48 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
   }
 }
 
+/// A fragment with only partial information that must not form a explicit
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:4301]:
```suggestion
/// A fragment with only partial information that must not form an explicit
```

File: packages/flutter/lib/src/rendering/object.dart [hunk 14]
```diff
@@ -4281,6 +4298,48 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
...
+  final SemanticsConfiguration config;
+
+  @override
+  void markAsExplicit() {
+    assert((){
+      throw FlutterError(
+        'SemanticsConfiguration created in SemanticsConfiguration.merger must '
+        'be not be explicit',
+      );
+    }());
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:4339]: This one throws a FlutterError inside an assert, addAll only has an assert. Why the difference?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 15]
```diff
@@ -1035,12 +1037,23 @@ class RenderParagraph extends RenderBox
...
+    for (final InlineSpanSemanticsInformation info in _semanticsInfo!) {
+      if (info.recognizer != null) {
+        needsAssembleSemanticsNode = true;
+        break;
+      }
+      needsChildConfigrationsDelegate = needsChildConfigrationsDelegate || info.isPlaceholder;
+    }
 
-    if (_semanticsInfo!.any((InlineSpanSemanticsInformation info) => info.recognizer != null)) {
+    if (needsAssembleSemanticsNode) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1050]: It's not clear to me why we can skip the delegate if there's a recognizer. Don't we still have the same problem with the placeholder nodes in this case?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 16]
```diff
@@ -1050,21 +1063,130 @@ class RenderParagraph extends RenderBox
...
+      final SemanticsConfiguration textSpanConfig = SemanticsConfiguration();
+      textSpanConfig.textDirection = textDirection;
+      textSpanConfig.attributedLabel = cachedStrings[textSpanLabelIndex];
+      builder.markAsMergeUp(textSpanConfig);
+    }
+    return builder.build();
+  }
+
+
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1189]: remove two of the extra new lines??

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 17]
```diff
@@ -1050,21 +1063,130 @@ class RenderParagraph extends RenderBox
...
   }
 
+  ChildSemanticsConfigurationsResult _childSemanticsConfigurationsDelegate(List<SemanticsConfiguration> childConfigs) {
+    if (_cachedAttributedLabels == null) {
+      return _buildChildSemanticsConfigurationsResultAndFillCache(childConfigs);
+    }
+    return _buildChildSemanticsConfigurationsResultFromCache(childConfigs);
+  }
+
+  ChildSemanticsConfigurationsResult _buildChildSemanticsConfigurationsResultAndFillCache(List<SemanticsConfiguration> childConfigs) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1090]: How are _buildChildSemanticsConfigurationsResultAndFillCache and _buildChildSemanticsConfigurationsResultAndFillCache kept in sync to make sure they end up producing the same ChildSemanticsConfigurationsResult?

File: packages/flutter/test/widgets/text_test.dart [hunk 18]
```diff
@@ -310,6 +310,156 @@ void main() {
...
+                baseline: TextBaseline.alphabetic,
+                child: SizedBox(width: 10.0),
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+    expect(tester.getSemantics(find.byType(Text)),
+        matchesSemantics(label: 'before \n mid\nfoo\n after'));
```
Comment (goderbauer) [inline packages/flutter/test/widgets/text_test.dart:379]: formatting, here and elsewhere in this file:

```suggestion
    expect(
      tester.getSemantics(find.byType(Text)),
      matchesSemantics(label: 'before \n mid\nfoo\n after'),
    );
```

File: packages/flutter/test/widgets/text_test.dart [hunk 19]
```diff
@@ -310,6 +310,156 @@ void main() {
...
+                child: Semantics(label: 'inner', container: true),
+              ),
+              const TextSpan(text: ' after'),
+            ],
+          ),
+        ),
+      ),
+    );
+    expect(tester.getSemantics(find.byType(Text)),
+        matchesSemantics(label: 'before \n after', children: <Matcher>[matchesSemantics(label: 'inner')]));
```
Comment (goderbauer) [inline packages/flutter/test/widgets/text_test.dart:452]: This one actually surprised me a little bit. I sort of expected there would be three sibling nodes: "before", "inner", and "after".

File: packages/flutter/test/widgets/text_test.dart [hunk 20]
```diff
@@ -310,6 +310,156 @@ void main() {
...
+
+    // If the Paragraph is not dirty it should use the cache correctly.
+    final RenderObject parent = tester.renderObject<RenderObject>(find.byType(Directionality));
+    parent.markNeedsSemanticsUpdate();
+    await tester.pumpAndSettle();
+
+    expect(tester.getSemantics(find.byType(Text)),
+        matchesSemantics(label: 'before \n after', children: <Matcher>[matchesSemantics(label: 'inner')]));
+  });
+
```
Comment (goderbauer) [inline packages/flutter/test/widgets/text_test.dart:462]: Maybe some tests that combines these cases with recognizers?

File: packages/flutter/lib/src/rendering/object.dart (hunk 14 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/object.dart:4339]: if it reaches addAll, it is something wrong with the framework, If it reaches here that means it is an user error.

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 15 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/paragraph.dart:1050]: If there is gesture recognizer, every continuous strings are separated into its own nodes so there won't be merging issue. and the assemblesemanticsnode is smart enough to insert each semanticsnode of widgetspans in the right places. So there is no need to use the childconfigurationdelegate API. As we discussed offline, if we do want to replace assemblesmeanticsnode with the childconfigurationdelegate API we need a way to set stable id and geometry

File: packages/flutter/test/widgets/text_test.dart (hunk 19 above)
Comment (chunhtai) [reply packages/flutter/test/widgets/text_test.dart:452]: I am not sure how we are going to support that, let's say we solve the unique id and geometry issue that synthetic SemanticsConfiguration can be explicit. How would it know when they should be explicit and not merge together. for example, how do they tell they need to merge in this case
```
[
              const TextSpan(text: 'before '),
              WidgetSpan(
                alignment: PlaceholderAlignment.baseline,
                baseline: TextBaseline.alphabetic,
                child: Semantics(label: 'inner'),
              ),
              const TextSpan(text: ' after'),
]
```
but not merge in this case

```
[
              const TextSpan(text: 'before '),
              WidgetSpan(
                alignment: PlaceholderAlignment.baseline,
                baseline: TextBaseline.alphabetic,
                child: Semantics(label: 'inner', container: true),
              ),
              const TextSpan(text: ' after'),
]
```

File: packages/flutter/test/widgets/text_test.dart (hunk 20 above)
Comment (chunhtai) [reply packages/flutter/test/widgets/text_test.dart:462]: I think we already have test that does that. The issue this pr trying to solve is that when there is only plain textspan and widgetspans that want to merge together.

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 17 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/paragraph.dart:1110]: added assert

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 15 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/paragraph.dart:1050]: Makes sense. Thanks for the explanation.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 21]
```diff
@@ -1050,21 +1063,148 @@ class RenderParagraph extends RenderBox
...
       config.textDirection = textDirection;
     }
   }
 
+  ChildSemanticsConfigurationsResult _childSemanticsConfigurationsDelegate(List<SemanticsConfiguration> childConfigs) {
+    if (_cachedAttributedLabels == null) {
+      final ChildSemanticsConfigurationsResult result = _buildChildSemanticsConfigurationsResultAndFillCache(childConfigs);
+      assert(() {
+        // Verify the cache will produce the same result.
+        final ChildSemanticsConfigurationsResult resultBuiltFromCache = _buildChildSemanticsConfigurationsResultAndFillCache(childConfigs);
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1088]: Should this be calling _buildChildSemanticsConfigurationsResultFromCache instead?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 22]
```diff
@@ -1050,21 +1063,148 @@ class RenderParagraph extends RenderBox
...
+            childConfigsIndex += 1;
+          } else {
+            assert(result.mergeUp[index].label == resultBuiltFromCache.mergeUp[index].label);
+          }
+        }
+        // All the ``childConfigs` must be included in the merge up.
+        assert(childConfigs.length == childConfigsIndex);
+        return true;
+      }());
+      return _buildChildSemanticsConfigurationsResultAndFillCache(childConfigs);
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1105]: Should this just return result instead of calling _buildChildSemanticsConfigurationsResultAndFillCache again?

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 17 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/paragraph.dart:1110]: This doesn't really address the maintainability problem. If we ever have to fix anything in this logic, one would have to do it twice. Plus, the (complicated) assert doesn't really inspire confidence that the two are actually equal. For example, in some cases it just assumes configs are equal when the labels match.

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/paragraph.dart:1110]: ah yes, I tried to combine this two functions in the previous commit, but the code looks like a trainwreck.

so I guess what you meant earlier is to make it more readable than to separate them out cache version and non-cache version?

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/paragraph.dart:1110]: Yes, sorry if I wasn't clear...

**Comment by goderbauer** [comment]
(triage fly by:) Looks like checks are failing on this one.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 23]
```diff
@@ -1050,21 +1063,77 @@ class RenderParagraph extends RenderBox
...
+    int attributedLabelCacheIndex = 0;
+    InlineSpanSemanticsInformation? seenTextInfo;
+    _cachedCombinedSemanticsInfos ??= combineSemanticsInfo(_semanticsInfo!);
+    for (final InlineSpanSemanticsInformation info in _cachedCombinedSemanticsInfos!) {
+      if (info.isPlaceholder) {
+        if (seenTextInfo != null) {
+          builder.markAsMergeUp(_createSemanticsConfigForTextInfo(seenTextInfo, attributedLabelCacheIndex));
+          attributedLabelCacheIndex += 1;
+        }
+        // Mark every childConfig belongs to this placeholder to merge up group.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1096]: belongs -> belonging

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 24]
```diff
@@ -1050,21 +1063,77 @@ class RenderParagraph extends RenderBox
...
+    return builder.build();
+  }
+
+  SemanticsConfiguration _createSemanticsConfigForTextInfo(InlineSpanSemanticsInformation textInfo, int cacheIndex) {
+    assert(!textInfo.requiresOwnNode);
+    final List<AttributedString> cachedStrings = _cachedAttributedLabels ??= <AttributedString>[];
+    assert(cacheIndex <= cachedStrings.length);
+    final bool hasCache = cacheIndex < cachedStrings.length;
+
+    late AttributedString attributedLabel;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart:1121]: nit: I think dart should allow you to mark this as final instead of late since every branch of the if below it assigns it exactly once.

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
internal test failure blocked on https://github.com/flutter/flutter/pull/120996
